nx/lib/nx/defn/evaluator.ex

```diff
@@ -54,6 +54,25 @@ defmodule Nx.Defn.Evaluator do
     eval(expr, state, cache)
   end
 
+  defp eval(
+         %Nx.Tensor{
+           data: %Expr{
+             op: :optional,
+             args: [expr, default_impl_expr]
+           }
+         },
+         state,
+         cache
+       ) do
+    {backend, _backend_options} = Nx.default_backend()
+
+    if function_exported?(backend, expr.data.op, length(expr.data.args)) do
+      eval_apply(expr.data.op, expr, state, cache)
+    else
+      eval_apply(default_impl_expr.data.op, default_impl_expr, state, cache)
+    end
```

Comment:
```suggestion
    %{data: %{op: op, args: args}} = expr

    if function_exported?(Nx.Shared.list_impl!(args), op, length(args)) do
      eval(expr, state, cache)
    else
      eval(default_impl_expr, state, cache)
    end
```

Reply:
Btw, you may want to move this whole clause to eval_apply. The difference is that the results of `eval_apply` are cached and we likely want to cache this so we don't lookup the backend several times? 🤔